src/java/org/apache/cassandra/config/DataStorageSpec.java

```diff
@@ -255,6 +219,210 @@ public String toString()
         return quantity + unit.symbol;
     }
 
+    /**
+     * Represents a data storage used for Cassandra configuration. The bound is [0; Long.MAX_VALUE) in bytes.
+     * If the user sets a different unit - we still validate that converted to bytes the quantity will not exceed
+     * that upper bound. (CASSANDRA-17571)
+     */
+    public static class LongBytesBound extends DataStorageSpec
```

Comment:
Note also that if we manage to get rid of the old properties in a future major release, and we always use long/bytes for all properties, we could remove the subclasses. If that happened we could remove all the subclasses of `DataStorageSpec` and keep the direct usages of the parent class.

Reply:
All valid points, that was the initial reason why we had it not abstract.
Now, after gathering more data and observing people's behavior in the past few months after the original patch landed I prefer the abstract class for a few reasons:
- People keep on adding new parameters with different unit because they are porting big features from their forks which are already tested and there was a reason to have parameter in a particular unit. So I don't think soon anything will be migrated to the smallest unit. 
- One issue is that in `Config` I noticed example of an easy to miss bug which can lead to problems. 
`paxos_purge_grace_period` - while it is set with smallest unit seconds, snakeyaml sees DurationSpec `paxos_purge_grace_period` and doesn't care about the constructor used in Config class. It looks at the default `DurationSpec` constructor and you are able to set nanoseconds and we are back to basics. And this was a bug introduced by a very senior community member which talks to me that we will see it more often than not.
Now, with the abstract DurationSpec what happens is - we don't get compile error and it is possible to have `DurationSpec paxos_purge_grace_period` but if someone tries to add in the yaml the parameter they will not be able to start the node:
 ` No single argument constructor found for class org.apache.cassandra.config.DurationSpec : null`
While annoying, I find it better preventive against us. First, I would hope people will test their parameters work if they are uncommented more often than testing them for min unit, second - it is less risky for users if they hit a bug and report it then if they start with something and get something totally unexpected. Example: getting the streaming to 0 instead of a very small value unthrottles it. 
Also, thinking more about it, I can try to actually add a test which writes a yaml from current Config class, streaming the properties and trying to start the node? This will help to be sure that no one tried to use the abstract class. WDYT?

Reply:
> I am thinking of probably updating cassandra.yaml and make a note for every property

I wouldn't hurt, although I guess I'd try to do it in a minimal/standard format like...

```
[min|smallest] unit: MiB
```

Reply:
```
Now, with the abstract DurationSpec what happens is - we don't get compile error and it is possible to have DurationSpec paxos_purge_grace_period but if someone tries to add in the yaml the parameter they will not be able to start the node
```
Yeah, this seems like the safer situation.

Reply:
Ok, let's see what Andres think about this and I can push the final version with the addressed comments on Monday. If he is fine with what I said, I can think of a test as the one suggested to help our future selves with this potential bug that can lead to null exception on startup

Reply:
That odd usage of `paxos_purge_grace_period` is convincing enough, let's keep the base class abstract.